### DISCUSSION TITLE
Trigger to execute script when a task gets assigned

### DISCUSSION THREAD

**Discussion by RolandvonBorstel** [sadieblue/agilityblue.workspace-scripts#8] [Q&A] [answered]
I would like to execute a script when a tasks switches from unassigned to assigned and with that from state "New" to "In Progress". Is there a trigger I can use to execute a script when this happens?

**Comment by mrabb24** [comment]
Hello, if you use a before save script on an update task event, you can use the get-task command to retrieve the task as it is currently saved. Then you would compare the status/assignment fields on the current task values compared to the ones the inbound-object task is about to save as. This should work for you in this scenario.

**Reply by RolandvonBorstel** [reply]
Hi Matt. Thanks for the suggestion. I will see if that works for me.

**Reply by RolandvonBorstel** [reply]
I have tried this method and it works fine when I edit a task or assign it using the buttons. But the script is preventing me from adding a form using the "Add Form" button. Is there any way to prevent this?

**Reply by mrabb24** [reply]
Does the execution results provide any error information or do you have a code snippet that causes this for me to look at?

**Comment by RolandvonBorstel** [comment]
![image](https://github.com/user-attachments/assets/23281e13-8ea6-406a-b613-4cd53400bf5c)

Here is a simple script I created to reproduce the error.

$InboundEntry    = Get-InboundObjectInstance
return $InboundEntry

All I did was to use the "Add Form" button on the task menu.

![image](https://github.com/user-attachments/assets/7462fee1-436e-4da8-9654-2446f06032b8)

**Comment by RolandvonBorstel** [comment]
HEre is the right error message:

![image](https://github.com/user-attachments/assets/26cb1cbf-1942-44f1-b4d7-72ca9a9bded1)

**Comment by RolandvonBorstel** [comment]
An additional error message returned from the script is: [Uncaught Exception] Cannot create and populate list type System.Linq.Enumerable+d__63`1[AgilityBlue.ViewModels.Workspace.TaskFormViewModel]. Path '', line 1, position 1.

**Reply by mrabb24** [reply]
Thank you, this is very helpful. I also getting this error in my test environment, though I'm not sure why. @bmward80, are you familiar with this behavior?

**Reply by bmward80** [reply]
No, but after doing some quick JSON tests, it appears that the act of adding forms is not passing in the `InboundObjectInstance` - it comes back as null. It looks like it is passing the task id just fine, but not the incoming forms that would be added to the task. I know that adding forms uses the `~/odata/workspace/tasks/{taskId}/addForms` API endpoint which only accepts a collection of forms instead of the full task, I am going to do some digging around and report back.

![image](https://github.com/user-attachments/assets/ceff66d9-107e-429b-b276-9ddd4d630f38)

Brandon

**Reply by bmward80** [reply, marked as answer]
This has been classified as a defect in the scripting pipeline specific to adding task forms. The system is sending an unserializable type to the scripting pipeline that is preventing it from being sent back. We're trying to see if there is any workaround within the scripts themself to allow this, but for the time being the only workaround is to add task forms by editing the task instead of using the Add Task Form button/menu item. I will keep you updated on any news, workarounds, or updates as it develops.

**Reply by RolandvonBorstel** [reply]
Understood. We will stay away from pre save task triggers then until this defect has been fixed.

**Reply by bmward80** [reply]
@RolandvonBorstel,

Some objects, most notably tasks, contain action information other than "created", "updated", and "deleted" and a different object is actually injected into the script other than the event object. You can think of these actions like "sub events". In this case, the act of adding forms will actually inject the forms list into your script instead of task because this particular action invokes the /addForms API endpoint action that takes in a list of forms instead of a task. You can use the special `$ABVAREventAction` variable to check what kind of action is occurring to handle various scenarios where different objects are injected for the same kind of event. Here's an example of how to handle the case with adding forms:

```powershell
$taskId = Get-InboundObjectId
$inboundEntry = Get-InboundObjectInstance

Write-Output "Event Action: $ABVAREventAction"

if ($ABVAREventAction -eq "ADD_FORMS") {
    Write-Output "Ignoring task $taskId because the user is just adding forms"
    return ,$inboundEntry
}

Write-Output "Updating task $taskId"

return $inboundEntry
```

Note that in the above, we're adding a comma before returning the `$inboundEntry` because PowerShell will convert a single-item array into a single object, and AB is expecting a list of forms back. Adding the comma before will force the object to be within an array and not doing this will cause the script to error and prevent the forms from being added.

There are other actions that you can consider for tasks that the `$ABVAREventAction` variable will provide:

```
CREATE, UPDATE, DELETE (provides/expects an instance of a task)
ASSIGN: (provides/expects an instance of a string - the id of the user)
ADD_FORMS: (provides/expects a list of task forms)
QUALITY_CONTROL, (provides/expects a QC status id and task status id object)
QUALITY_CONTROL, DONE (provides/expects the task id)
```

We're going to work on a write-up on the help docs about these event actions that should help clear things up further for before-save actions. I hope this has been helpful, but let me know if you have any other questions

Brandon

**Reply by RolandvonBorstel** [reply]
Thanks for the elaborate answer. I have tried this code and it addresses the add form action. Unfortunately, it fails for the un-assign action.  I am getting this error:

System.AggregateException: One or more errors occurred. (Object reference not set to an instance of an object.)
 ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at AgilityBlue.ViewModels.KeyContainer`1.ToString() in /dist/src/business/agilityblue.view-models/KeyContainer.cs:line 22
   at AgilityBlue.WebApi.Scripting.Controllers.ScriptsController.<>c__DisplayClass2_0.b__0() in /dist/src/presentation/agilityblue.web-api.scripting/Controllers/ScriptsController.cs:line 110
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at AgilityBlue.WebApi.Scripting.Controllers.ScriptsController.Execute(ExecuteWorkspaceScriptCommand request, Boolean includeOutput) in /dist/src/presentation/agilityblue.web-api.scripting/Controllers/ScriptsController.cs:line 12

**Reply by bmward80** [reply]
Hi @RolandvonBorstel, I've been advised that this is an OK error to ignore as the act of unassigning should still occur. The underlying issue here is that the system is trying to call a "to string" type of method on the assignee id, but since it's null because unassigning someone causes it to be null, it fails at doing that. Devs will address this and catch the error in a future update to the view models library.

Thank you
Brandon
